File: src/main/java8/org/apache/beam/examples/complete/game/solutions/Exercise5.java
/*
 * Copyright (C) 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.apache.beam.examples.complete.game.solutions;

import com.google.api.services.bigquery.model.TableFieldSchema;
import com.google.api.services.bigquery.model.TableReference;
import com.google.api.services.bigquery.model.TableRow;
import com.google.api.services.bigquery.model.TableSchema;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.apache.beam.examples.complete.game.utils.GameEvent;
import org.apache.beam.examples.complete.game.utils.Options;
import org.apache.beam.runners.dataflow.DataflowRunner;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.PipelineResult;
import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.CreateDisposition;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.WriteDisposition;
import org.apache.beam.sdk.metrics.Counter;
import org.apache.beam.sdk.metrics.Metrics;
import org.apache.beam.sdk.options.Default;
import org.apache.beam.sdk.options.Description;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.options.StreamingOptions;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.MapElements;
import org.apache.beam.sdk.transforms.Mean;
import org.apache.beam.sdk.transforms.PTransform;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.Sum;
import org.apache.beam.sdk.transforms.Values;
import org.apache.beam.sdk.transforms.View;
import org.apache.beam.sdk.transforms.windowing.FixedWindows;
import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
import org.apache.beam.sdk.transforms.windowing.Window;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollection;
import org.apache.beam.sdk.values.PCollectionView;
import org.apache.beam.sdk.values.TypeDescriptors;
import org.joda.time.Duration;
import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Fifth in a series of coding exercises in a gaming domain.
 *
 * <p>This exercise introduces side inputs.
 *
 * <p>See README.md for details.
 */
public class Exercise5 {

  private static final Logger LOG = LoggerFactory.getLogger(Exercise5.class);

  /**
   * Filter out all but those users with a high clickrate, which we will consider as 'spammy' users.
   * We do this by finding the mean total score per user, then using that information as a side
   * input to filter out all but those user scores that are > (mean * SCORE_WEIGHT)
   */
  public static class CalculateSpammyUsers
      extends PTransform<PCollection<KV<String, Integer>>, PCollection<KV<String, Integer>>> {

    private static final Logger LOG = LoggerFactory.getLogger(CalculateSpammyUsers.class);
    private static final double SCORE_WEIGHT = 2.5;

    @Override
    public PCollection<KV<String, Integer>> expand(PCollection<KV<String, Integer>> userScores) {

      // Get the sum of scores for each user.
      PCollection<KV<String, Integer>> sumScores =
          userScores.apply("UserSum", Sum.<String>integersPerKey());

      // Extract the score from each element, and use it to find the global mean.
      final PCollectionView<Double> globalMeanScore =
          sumScores
              .apply(Values.<Integer>create())
              .apply(Mean.<Integer>globally().asSingletonView());

      // Filter the user sums using the global mean.
      PCollection<KV<String, Integer>> filtered =
          sumScores.apply("ProcessAndFilter",
              ParDo
                  // use the derived mean total score as a side input
                  .of(
                      new DoFn<KV<String, Integer>, KV<String, Integer>>() {
                        private final Counter numSpammerUsers = Metrics
                            .counter("main", "SpammerUsers");

                        @ProcessElement
                        public void processElement(ProcessContext c) {
                          Integer score = c.element().getValue();
                          Double gmc = c.sideInput(globalMeanScore);
                          if (score > (gmc * SCORE_WEIGHT)) {
                            LOG.info(
                                "user "
                                    + c.element().getKey()
                                    + " spammer score "
                                    + score
                                    + " with mean "
                                    + gmc);
                            numSpammerUsers.inc();
                            c.output(c.element());
                          }
                        }
                      })
                  .withSideInputs(globalMeanScore));
      return filtered;
    }
  }

  /**
   * Calculate and output an element's session duration.
   */
  private static class UserSessionInfoFn extends DoFn<KV<String, Integer>, Integer> {

    @ProcessElement
    public void processElement(ProcessContext c, IntervalWindow w) {
      int duration = new Duration(w.start(), w.end()).toPeriod().toStandardMinutes().getMinutes();
      c.output(duration);
    }
  }

  /**
   * Options supported by {@link Exercise5}.
   */
  interface Exercise5Options extends Options, StreamingOptions {

    @Description("Numeric value of fixed window duration for user analysis, in minutes")
    @Default.Integer(5)
    Integer getFixedWindowDuration();

    void setFixedWindowDuration(Integer value);
  }

  public static void main(String[] args) throws Exception {

    Exercise5Options options =
        PipelineOptionsFactory.fromArgs(args).withValidation().as(Exercise5Options.class);
    // Enforce that this pipeline is always run in streaming mode.
    options.setStreaming(true);
    options.setRunner(DataflowRunner.class);
    Pipeline pipeline = Pipeline.create(options);

    TableReference teamTable = new TableReference();
    teamTable.setDatasetId(options.getOutputDataset());
    teamTable.setProjectId(options.as(GcpOptions.class).getProject());
    teamTable.setTableId(options.getOutputTableName());

    PCollection<GameEvent> rawEvents = pipeline.apply(new Exercise3.ReadGameEvents(options));

    // Extract username/score pairs from the event stream
    PCollection<KV<String, Integer>> userEvents =
        rawEvents.apply(
            "ExtractUserScore",
            MapElements
                .into(TypeDescriptors.kvs(TypeDescriptors.strings(), TypeDescriptors.integers()))
                .via((GameEvent gInfo) -> KV.<String, Integer>of(gInfo.getUser(),
                    gInfo.getScore())));

    // Calculate the total score per user over fixed windows, and
    // cumulative updates for late data.
    final PCollectionView<Map<String, Integer>> spammersView =
        userEvents
            .apply("FixedWindowsUser",
                Window.<KV<String, Integer>>into(
                        FixedWindows.of(
                            Duration.standardMinutes(options.getFixedWindowDuration()))))

            // Filter out everyone but those with (SCORE_WEIGHT * avg) clickrate.
            // These might be robots/spammers.
            .apply("CalculateSpammyUsers", new CalculateSpammyUsers())
            // Derive a view from the collection of spammer users. It will be used as a side input
            // in calculating the team score sums, below.
            .apply("CreateSpammersView", View.<String, Integer>asMap());

    // Calculate the total score per team over fixed windows,
    // and emit cumulative updates for late data. Uses the side input derived above-- the set of
    // suspected robots-- to filter out scores from those users from the sum.
    // Write the results to BigQuery.
    rawEvents
        .apply("WindowIntoFixedWindows",
            Window
                .<GameEvent>into(
                    FixedWindows.of(Duration.standardMinutes(options.getFixedWindowDuration()))))
        // Filter out the detected spammer users, using the side input derived above.
        .apply("FilterOutSpammers",
            ParDo
                .of(
                    new DoFn<GameEvent, GameEvent>() {
                      @ProcessElement
                      public void processElement(ProcessContext c) {
                        // If the user is not in the spammers Map, output the data element.
                        if (c.sideInput(spammersView).get(c.element().getUser().trim()) == null) {
                          c.output(c.element());
                        }
                      }
                    })
                .withSideInputs(spammersView))
        // Extract and sum teamname/score pairs from the event data.
        .apply("ExtractTeamScore", new Exercise1.ExtractAndSumScore("team"))
        // Write the result to BigQuery
        .apply("FormatTeamWindows", ParDo.of(new FormatTeamWindowFn()))
        .apply(
            BigQueryIO.writeTableRows().to(teamTable)
                .withSchema(FormatTeamWindowFn.getSchema())
                .withCreateDisposition(CreateDisposition.CREATE_IF_NEEDED)
                .withWriteDisposition(WriteDisposition.WRITE_APPEND));

    PipelineResult result = pipeline.run();
    result.waitUntilFinish();
  }

  /**
   * Format a KV of team and associated properties to a BigQuery TableRow.
   */
  protected static class FormatTeamWindowFn extends DoFn<KV<String, Integer>, TableRow> {

    @ProcessElement
    public void processElement(ProcessContext c, IntervalWindow window) {
      TableRow row =
          new TableRow()
              .set("team", c.element().getKey())
              .set("total_score", c.element().getValue())
              .set("window_start", window.start().getMillis() / 1000)
              .set("processing_time", Instant.now().getMillis() / 1000);
      c.output(row);
    }

    static TableSchema getSchema() {
      List<TableFieldSchema> fields = new ArrayList<>();
      fields.add(new TableFieldSchema().setName("team").setType("STRING"));
      fields.add(new TableFieldSchema().setName("total_score").setType("INTEGER"));
      fields.add(new TableFieldSchema().setName("window_start").setType("TIMESTAMP"));
      fields.add(new TableFieldSchema().setName("processing_time").setType("TIMESTAMP"));
      return new TableSchema().setFields(fields);
    }
  }
}

File: py/solutions/exercise5.py
# Filter 'cheating' or 'spammy' users from the game results.
# Computes the global mean score and filters users that are 
# some threshold above that score.
from __future__ import absolute_import

import logging
import re
import time

import apache_beam as beam
from apache_beam.io import ReadFromText
from apache_beam.io import ReadFromPubSub
from apache_beam.io import WriteToText
from apache_beam.metrics import Metrics
from apache_beam.metrics.metric import MetricsFilter
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.options.pipeline_options import SetupOptions
from apache_beam.options.pipeline_options import StandardOptions
from apache_beam.options.pipeline_options import GoogleCloudOptions
from apache_beam.transforms import trigger
from util.util import GameEvent
from util.util import ParseEvent
from util.util import ParseEventFn
from util.util import ParseArgs
import apache_beam.transforms.window as window

# Defines the BigQuery schemas.
USER_SCHEMA = ('user:STRING,'
               'total_score:INTEGER,'
               'processing_time:TIMESTAMP')
TEAM_SCHEMA = ('team:STRING,' 'total_score:INTEGER,' 'window_start:TIMESTAMP')


class ExtractAndSumScore(beam.PTransform):
    def __init__(self, field):
        super(ExtractAndSumScore, self).__init__()
        self.field = field

    def expand(self, p):
        return (p
                | 'extract_field' >> beam.Map(
                    lambda x: (vars(x)[self.field], x.score))
                | beam.CombinePerKey(sum)
                )


class WindowedUserScores(beam.PTransform):
    """Extract user/score pairs via in fixed windows."""
    def __init__(self, duration):
        super(WindowedUserScores, self).__init__()
        self.duration = duration

    def expand(self, p):
        return (p
                | 'window' >> beam.WindowInto(
                    window.FixedWindows(self.duration))
                | 'extract_user_score' >> ExtractAndSumScore('user')
                )


class FilterUser(beam.DoFn):
    """Filter a user if their score * score_weight > avg_score."""
    def __init__(self, score_weight):
        super(FilterUser, self).__init__()
        self.score_weight = score_weight
        self.num_spammy_users = Metrics.counter(self.__class__,
                                                'num_spammy_users')

    def process(self, user_score, avg_score=beam.DoFn.SideInputParam):
        user, score = user_score
        if score * self.score_weight > avg_score:
            logging.error('User %s filtered as spammy', user)
            self.num_spammy_users.inc()
            yield user


class ComputeSpammyUsers(beam.PTransform):
    """Compute users with a high clickrate, which we will consider spammy.
     We do this by finding the mean total score per user and filter out
     those with scores that are greater than the mean * score_weight
    """
    def __init__(self, score_weight):
        super(ComputeSpammyUsers, self).__init__()
        self.score_weight = score_weight

    def expand(self, p):
        avg_score = (p
                | beam.Values()
                | beam.CombineGlobally(
                    beam.combiners.MeanCombineFn()).as_singleton_view()
                )
        return (p
                | 'compute_spammers' >> beam.ParDo(
                    FilterUser(self.score_weight), avg_score=avg_score)
                )


class FilterSpammers(beam.DoFn):
    """Remove users found in the spam list."""
    def __init__(self):
        super(FilterSpammers, self).__init__()
        self.filtered_scores = Metrics.counter(self.__class__,
                                               'filtered_scores')

    def process(self, elem, spammers=beam.DoFn.SideInputParam):
        user = elem.user
        if user not in spammers:
            yield elem
        else:
            self.filtered_scores.inc()


class WindowedTeamScore(beam.PTransform):
    """Calculates scores for each team within the configured window duration"""
    def __init__(self, duration, spammers):
        super(WindowedTeamScore, self).__init__()
        self.duration = duration
        self.spammers = spammers

    def expand(self, p):
        return (p
                | 'window' >> beam.WindowInto(
                    window.FixedWindows(self.duration))
                | 'filter_spammers' >> beam.ParDo(
                    FilterSpammers(), spammers=self.spammers)
                | 'extract_team_score' >> ExtractAndSumScore('team')
                )


class FormatTeamScoreSum(beam.DoFn):
    def process(self, team_score, window=beam.DoFn.WindowParam):
        team, score = team_score
        start = int(window.start)
        yield {
            'team': team,
            'total_score': score,
            'window_start': start,
        }


class FormatUserScoreSum(beam.DoFn):
    def process(self, user_score, window=beam.DoFn.WindowParam):
        user, score = user_score
        yield {
            'user': user,
            'total_score': score,
            'processing_time': time.time(),
        }


def Run(argv=None):
    known_args, pipeline_args = ParseArgs(argv)
    pipeline_options = PipelineOptions(pipeline_args)
    pipeline_options.view_as(SetupOptions).save_main_session = True
    p = beam.Pipeline(options=pipeline_options)
    window_duration = 1 * 60  # 1 minute windows.
    if known_args.topic:
        pipeline_options.view_as(StandardOptions).streaming = True

    project = pipeline_options.view_as(GoogleCloudOptions).project
    timestamp_attribute = 'timestamp_ms'
    events = None
    if (not known_args.topic):
        events = (p
                | 'read' >> ReadFromText(known_args.input)
                | 'parse' >> beam.FlatMap(ParseEventFn())
                | 'add_event_timestamps' >> beam.Map(
                    lambda x: beam.window.TimestampedValue(x, x.timestamp)))
    else:
        events = (p
                | 'read' >> ReadFromPubSub(
                    topic=known_args.topic,
                    timestamp_attribute='timestamp_ms')
                | 'decode' >> beam.ParDo(ParseEventFn()))

    user_scores = (events
            | 'window_user_scores' >> WindowedUserScores(window_duration))
    spammers = beam.pvalue.AsList(user_scores
            | 'compute_spammers' >> ComputeSpammyUsers(2.5))

    _ = (events
         | 'windowed_team_score' >> WindowedTeamScore(window_duration, spammers)
         | 'format_team_score_sum' >> beam.ParDo(FormatTeamScoreSum())
         | 'write_teams_to_bigquery' >> beam.io.WriteToBigQuery(
             known_args.output_tablename, known_args.output_dataset, project,
             TEAM_SCHEMA)
         )

    p.run().wait_until_finish()


if __name__ == '__main__':
    logging.getLogger().setLevel(logging.INFO)
    Run()
